keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/LogProbGradientCalculator.java

```diff
@@ -0,0 +1,160 @@
+package io.improbable.keanu.vertices.dbl.nonprobabilistic.diff;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.google.common.base.Preconditions;
+
+import io.improbable.keanu.network.LambdaSection;
+import io.improbable.keanu.tensor.Tensor;
+import io.improbable.keanu.tensor.TensorShape;
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import io.improbable.keanu.vertices.Probabilistic;
+import io.improbable.keanu.vertices.Vertex;
+import io.improbable.keanu.vertices.VertexId;
+import io.improbable.keanu.vertices.dbl.Differentiator;
+import io.improbable.keanu.vertices.dbl.DoubleVertex;
+
+public class LogProbGradientCalculator {
+
+    private final Set<? extends Vertex<?>> logProbOfVertices;
+    private final Set<? extends Vertex<?>> wrtVertices;
+
+    private final Map<Vertex, Set<DoubleVertex>> parentToLatentLookup;
+    private final Map<Vertex, Set<DoubleVertex>> verticesWithNonzeroDiffWrtLatent;
+
+    public LogProbGradientCalculator(List<? extends Vertex> logProbOfVerticesList, List<? extends Vertex<?>> wrtVerticesList) {
+        this.logProbOfVertices = new HashSet<>((List<Vertex<?>>) logProbOfVerticesList);
+        this.wrtVertices = new HashSet<>(wrtVerticesList);
+
+        parentToLatentLookup = getParentsThatAreConnectedToWrtVertices(logProbOfVertices);
+        verticesWithNonzeroDiffWrtLatent = getVerticesWithNonzeroDiffWrt(logProbOfVertices, parentToLatentLookup);
+    }
+
+    /**
+     * @return the partial derivatives with respect to a give set of latent vertices
+     */
+    public Map<VertexId, DoubleTensor> getJointLogProbGradientWrtLatents() {
```

Comment:
This seems to be one of the few methods remaining that has a `VertexId > DoubleTensor` mapping. Can we swap this out for `Vertex` to make it more consistent or is that too deeply nestled in `PartialDerivatives`?

Reply:
It's in the PartialDerivatives so that you don't need to have the entire network in order to have the gradient of that network. If the gradient calculation was distributed then it'd really come in handy.